### PR TITLE
Integrated thread pool, fixed all UTs and ITs

### PR DESCRIPTION
**Pull request by just-now**

### CODE REVIEW

**Comment by just-now** [comment]
WIP

File: Makefile.am [hunk 1]
```diff
@@ -15,7 +15,7 @@ endif
 if DEBUG_ENABLED
   AM_CFLAGS += -g3
 else
-  AM_CFLAGS += -O2
+  AM_CFLAGS += -O0
```
[the remark this answers is not in the document]
Comment (just-now) [reply Makefile.am:18]: added by mistake. fixed.

File: src/gateway.c [hunk 2]
```diff
@@ -514,17 +518,23 @@ static int handle_exec(struct gateway *g, struct handle *req)
...
 	sqlite3_stmt *stmt = req->stmt;
 	assert(stmt != NULL);
 	struct response_rows response;
 	int rc;
 
-	rc = query__batch(stmt, req->buffer);
+	if (half == POOL_TOP_HALF) {
+		req->rc = query__batch(stmt, req->buffer);
+		return;
+	} // else POOL_TOP_HALF =>
```
[the remark this answers is not in the document]
Comment (just-now) [reply src/gateway.c:535]: fixed

File: src/gateway.c [hunk 3]
```diff
@@ -514,17 +518,23 @@ static int handle_exec(struct gateway *g, struct handle *req)
  * given request with a single batch of rows.
  *
  * A single batch of rows is typically about the size of a memory page. */
-static void query_batch(struct gateway *g)
+static void query_batch2(struct handle *req, int half)
```
[the remark this answers is not in the document]
Comment (just-now) [reply src/gateway.c:521]: fixed to `query_batch_async()`

File: src/lib/threadpool.h [hunk 4]
```diff
@@ -70,10 +70,18 @@ struct pool_work_s
...
 };
 
 enum {
 	POOL_QOS_PRIO_FAIR = 2,
+
+	POOL_TOP_HALF = 0x109,
+	POOL_BOTTOM_HALF = 0xb01103,
+
+	POOL_FOR_UT_NON_CLEAN_FINI = 1u << 0,
+	POOL_FOR_UT_NOT_ASYNC      = 1u << 1,
```
[the remark this answers is not in the document]
Comment (just-now) [reply src/lib/threadpool.h:83]: not intended. 
The with `POOL_QOS_PRIO_FAIR` and `POOL_FOR_UT_NON_ASYNC` is just a set of unrelated constants.

File: src/server.c [hunk 5]
```diff
@@ -81,17 +82,25 @@ int dqlite__init(struct dqlite_node *d,
...
+
 	rv = uv_loop_init(&d->loop);
 	if (rv != 0) {
 		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE,
 			 "uv_loop_init(): %s", uv_strerror(rv));
 		rv = DQLITE_ERROR;
 		goto err_after_vfs_init;
 	}
-	rv = raftProxyInit(&d->raft_transport, &d->loop);
+	rv = pool_init(&d->pool, &d->loop, 4, POOL_QOS_PRIO_FAIR);
```
[the remark this answers is not in the document]
Comment (just-now) [reply src/server.c:93]: Please tell me more. Where to take the number of threads from? Shall I put it into `struct config`? Is there any configuration file?

File: test/integration/test_client.c [hunk 6]
```diff
@@ -116,9 +120,9 @@ TEST(client, query, setUp, tearDown, 0, client_params)
 	EXEC(stmt_id, &last_insert_id, &rows_affected);
 
 	PREPARE("SELECT n FROM test", &stmt_id);
-	QUERY(stmt_id, &rows);
+	QUERY(stmt_id, &f->rows);
 
-	clientCloseRows(&rows);
```
[the remark this answers is not in the document]
Comment (just-now) [reply test/integration/test_client.c:121]: Yes. I do understand.

At this point in the code the pool hasn't yet consumed rows. That's why the finalization part was put after the point of time when the rows have been consumed by the pool.

Before the pool was integrated, calls above line 121 were consuming rows in the same thread. That's why it's safe to call clientCloseRows() here at 121.

File: test/unit/test_conn.c [hunk 7]
```diff
@@ -347,6 +354,9 @@ TEST_CASE(exec, close_while_in_flight, NULL)
 
 	test_uv_run(&f->loop, 1);
 
+	/* TODO: This needs an additional long comment. */
```
[the remark this answers is not in the document]
Comment (just-now) [reply test/unit/test_conn.c:357]: Yes, it's worth putting a comment near by `POOL_FOR_UT_NON_CLEAN_FINI` def.

The comment is:
Setting POOL_FOR_UT_NON_CLEAN_FINI relaxes pool's invariant during the finalization w.r.t. to pass a few tests checking failures with non-clean unit-test termination.

File: src/gateway.c [hunk 8]
```diff
@@ -548,6 +558,31 @@ static void query_batch(struct gateway *g)
 	}
 }
 
+static void qb_top(pool_work_t *w)
+{
+	struct handle *req = CONTAINER_OF(w, struct handle, work);
+	query_batch2(req, POOL_TOP_HALF);
```
[the remark this answers is not in the document]
Comment (just-now) [reply src/gateway.c:564]: This is an interface adapter. As `pool_queue_work()` expects `work_cb()` and `after_work_cb()`. Therefore should be either two functions for this or one function with an appropriate state to distinguish whether it is a top or bottom half.

Anyways it'd be interesting to look at alternatives.

**Comment by cole-miller** [comment]
@MathieuBordere yeah agreed about the integration branch, we want to be able to cut releases that use the old code until the new stuff is ready for prime time

File: src/server.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (cole-miller) [reply src/server.c:93]: Yes I would put this in `struct config`, with a default value and a public setter (`dqlite_node_set_thread_count` or something)

**Comment by cole-miller** [comment]
I've created the `dqlite-next` branch, we can use it as a target for PRs like this.

**Pull request closed without merging**